# **Part 2. Theory**

## **1. Involutory Keys in the Shift Cipher**

![Shift Cipher](imgs/shift_cipher.jpg)

## **2. Permutation Cipher**

### **a) Cálculo de la permutación inversa \( \pi^{-1} \)**
![PI Cipher](imgs/a_pi.jpg)

### **b) Descifrado del texto cifrado**
![DEC Cipher](imgs/b_decifrado.jpg)

## **3. Below are given four examples of ciphertext, one obtained from a Substitution Cipher, one from a Vigenere Cipher, one from an Affine Cipher, and one unspecified. In each case, the task is to determine the plaintext. Give a clearly written description of the steps you followed to decrypt each ciphertext. This should include all statistical analysis and computations you performed.**

## - **Substitution:**

EMGLOSUDCGDNCUSWYSFHNSFCYKDPUMLWGYICOXYSIPJCK
QPKUGKMGOLICGINCGACKSNISACYKZSCKXECJCKSHYSXCG
OIDPKZCNKSHICGIWYGKKGKGOLDSILKGOIUSIGLEDSPWZU
GFZCCNDGYYSFUSZCNXEOJNCGYEOWEUPXEZGACGNFGLKNS
ACIGOIYCKXCJUCIUZCFZCCNDGYYSFEUEKUZCSOCFZCCNC
IACZEJNCSHFZEJZEGMXCYHCJUMGKUCY

**Solution** 

![Probabilitie Table](imgs/tabla_prob.png)

In [2]:
import collections

# Encrypted text (ciphertext)
ciphertext = """
EMGLOSUDCGDNCUSWYSFHNSFCYKDPUMLWGYICOXYSIPJCK
QPKUGKMGOLICGINCGACKSNISACYKZSCKXECJCKSHYSXCG
OIDPKZCNKSHICGIWYGKKGKGOLDSILKGOIUSIGLEDSPWZU
GFZCCNDGYYSFUSZCNXEOJNCGYEOWEUPXEZGACGNFGLKNS
ACIGOIYCKXCJUCIUZCFZCCNDGYYSFEUEKUZCSOCFZCCNC
IACZEJNCSHFZEJZEGMXCYHCJUMGKUCY
""".replace("\n", "")

# Count letter frequency
letter_freq = collections.Counter(ciphertext)

# Filter only alphabetic letters and sort by frequency
sorted_letter_freq = sorted(letter_freq.items(), key=lambda item: item[1], reverse=True)

# Count trigrams (sequences of 3 consecutive letters)
trigram_freq = collections.Counter(ciphertext[i:i+3] for i in range(len(ciphertext) - 2))

# Sort trigrams by frequency
sorted_trigram_freq = sorted(trigram_freq.items(), key=lambda item: item[1], reverse=True)

# Display letter frequencies
print("Letter | Frequency")
print("-----------------")
for letter, freq in sorted_letter_freq:
    print(f"   {letter}   |     {freq}")

# Display trigram frequencies
print("\nTrigram | Frequency")
print("-------------------")
for trigram, freq in sorted_trigram_freq[:10]:  # Show only top 10 trigrams
    print(f"  {trigram}  |    {freq}")


Letter | Frequency
-----------------
   C   |     37
   G   |     24
   S   |     20
   K   |     18
   Y   |     15
   I   |     15
   U   |     14
   N   |     13
   Z   |     13
   E   |     12
   O   |     10
   F   |     9
   D   |     8
   L   |     7
   X   |     7
   J   |     7
   P   |     6
   M   |     5
   W   |     5
   H   |     5
   A   |     5
   Q   |     1

Trigram | Frequency
-------------------
  YSF  |    3
  GOI  |    3
  FZC  |    3
  ZCC  |    3
  CCN  |    3
  CYK  |    2
  JCK  |    2
  GOL  |    2
  ICG  |    2
  CGI  |    2


First, the most frequent letter in the ciphertext is assigned to 'E', as it is the most common letter in English. Then, additional mappings are applied based on observed digrams and trigrams, such as '0' → 'N', 'U' → 'T', since these combinations frequently appear in common words. As more correct substitutions are identified, new letters can be inferred from the structure of the resulting words. The code displays the frequency of each letter in the ciphertext, its ranking, and its assigned counterpart in English. Finally, a preliminary decrypted text is printed, allowing an evaluation of whether the substitutions follow a logical pattern and refining them as needed. Also, searching patterns such as CG - EA. Also, the triagrams with C at the end makes us assume is THE, as its the most use.

In [3]:
import collections

# Encrypted text (ciphertext)
ciphertext = """
EMGLOSUDCGDNCUSWYSFHNSFCYKDPUMLWGYICOXYSIPJCK
QPKUGKMGOLICGINCGACKSNISACYKZSCKXECJCKSHYSXCG
OIDPKZCNKSHICGIWYGKKGKGOLDSILKGOIUSIGLEDSPWZU
GFZCCNDGYYSFUSZCNXEOJNCGYEOWEUPXEZGACGNFGLKNS
ACIGOIYCKXCJUCIUZCFZCCNDGYYSFEUEKUZCSOCFZCCNC
IACZEJNCSHFZEJZEGMXCYHCJUMGKUCY
""".replace("\n", "")

# Count the frequency of each letter in the ciphertext
letter_freq = collections.Counter(ciphertext)

# Filter only alphabetic letters and sort by frequency
sorted_letter_freq = sorted(letter_freq.items(), key=lambda item: item[1], reverse=True)

# Create a ranking of letters based on frequency
rank = {sorted_letter_freq[i][0]: i + 1 for i in range(len(sorted_letter_freq))}

# Assign the most frequent letter to 'E' and continue with other substitutions based on analysis
substitution_map = {}
if sorted_letter_freq:
    mappings_from_analysis = {
        'C': 'E', 'Q': 'J', 'Z': 'H', 'N': 'H', 'U': 'T', 'S': 'O', 'O': 'N', 'K': 'S',
        'I': 'D', 'W': 'G', 'L': 'Y', 'X': 'P', 'J': 'C', 'E': 'I', 'P': 'U',
        'D': 'B', 'M': 'M'
    }
    substitution_map.update(mappings_from_analysis)

# Apply substitution
decrypted_text = "".join(substitution_map.get(c, c) for c in ciphertext)

# Display letter frequencies, ranking, and initial mapping
print("Letter | Frequency | Rank | Mapping")
print("--------------------------------------")
for letter, freq in sorted_letter_freq:
    mapped = substitution_map.get(letter, "?")
    print(f"  {letter}   |     {freq}      |  {rank[letter]}  |  {mapped}")

# Display the preliminary decrypted text without line breaks
print("\nPreliminary decrypted text:")
print(decrypted_text, end="\n\n")

Letter | Frequency | Rank | Mapping
--------------------------------------
  C   |     37      |  1  |  E
  G   |     24      |  2  |  ?
  S   |     20      |  3  |  O
  K   |     18      |  4  |  S
  Y   |     15      |  5  |  ?
  I   |     15      |  6  |  D
  U   |     14      |  7  |  T
  N   |     13      |  8  |  H
  Z   |     13      |  9  |  H
  E   |     12      |  10  |  I
  O   |     10      |  11  |  N
  F   |     9      |  12  |  ?
  D   |     8      |  13  |  B
  L   |     7      |  14  |  Y
  X   |     7      |  15  |  P
  J   |     7      |  16  |  C
  P   |     6      |  17  |  U
  M   |     5      |  18  |  M
  W   |     5      |  19  |  G
  H   |     5      |  20  |  ?
  A   |     5      |  21  |  ?
  Q   |     1      |  22  |  J

Preliminary decrypted text:
IMGYNOTBEGBHETOGYOFHHOFEYSBUTMYGGYDENPYODUCESJUSTGSMGNYDEGDHEGAESOHDOAEYSHOESPIECESOHYOPEGNDBUSHEHSOHDEGDGYGSSGSGNYBODYSGNDTODGYIBOUGHTGFHEEHBGYYOFTOHEHPINCHEGYINGITUPIHGAEGHFGYSHOAEDGNDYESPECTEDTHEFHEEHBGYYOFITI

The approach used in this refinement follows the principles outlined in the book for substitution cipher decryption. Initially, the ciphertext was processed using frequency analysis, identifying the most common letter and mapping it to 'E', the most frequent letter in English. Then, additional mappings were derived from digram and trigram patterns, refining the substitution.

However, the preliminary decryption still contained partially deciphered words, requiring refinements. To improve accurace we identify common English words that appeared partially recognizable in the decrypted text.
Matching word structures with expected patterns based on previously substituted letters.
Using regular expressions and predefined mappings to replace incomplete fragments with their correct English equivalents.
For instance:

"IMGY" was recognized as "I MAY", aligning with the structure of common phrases.
"BEGBLE" was corrected to "BE ABLE". 

By refining the text through pattern-based substitutions, we achieved a more coherent and readable decrypted message by changing letters to form these words. 

In [4]:
import collections

# Encrypted text (ciphertext)
ciphertext = """
EMGLOSUDCGDNCUSWYSFHNSFCYKDPUMLWGYICOXYSIPJCK
QPKUGKMGOLICGINCGACKSNISACYKZSCKXECJCKSHYSXCG
OIDPKZCNKSHICGIWYGKKGKGOLDSILKGOIUSIGLEDSPWZU
GFZCCNDGYYSFUSZCNXEOJNCGYEOWEUPXEZGACGNFGLKNS
ACIGOIYCKXCJUCIUZCFZCCNDGYYSFEUEKUZCSOCFZCCNC
IACZEJNCSHFZEJZEGMXCYHCJUMGKUCY
""".replace("\n", "")

# Count the frequency of each letter in the ciphertext
letter_freq = collections.Counter(ciphertext)

# Filter only alphabetic letters and sort by frequency
sorted_letter_freq = sorted(letter_freq.items(), key=lambda item: item[1], reverse=True)

# Create a ranking of letters based on frequency
rank = {sorted_letter_freq[i][0]: i + 1 for i in range(len(sorted_letter_freq))}

# Assign the most frequent letter to 'E' and continue with other substitutions based on analysis
substitution_map = {
    'C': 'E', 'Q': 'J', 'Z': 'H', 'N': 'L', 'U': 'T', 'S': 'O', 'O': 'N', 'K': 'S',
    'I': 'D', 'A': 'V', 'W': 'G', 'L': 'Y', 'X': 'P', 'J': 'C', 'E': 'I', 'P': 'U',
    'D': 'B', 'M': 'M', 'B': 'E', 'F': 'W', 'G': 'A', 'H': 'F', 'R': 'P', 'T': 'Y', 'Y': 'R'
}

# Apply substitution
decrypted_text = "".join(substitution_map.get(c, c) for c in ciphertext)

# Display letter frequencies, ranking, and initial mapping
print("Letter | Frequency | Rank | Mapping")
print("--------------------------------------")
for letter, freq in sorted_letter_freq:
    mapped = substitution_map.get(letter, "?")
    print(f"  {letter}   |     {freq}      |  {rank[letter]}  |  {mapped}")

# Display the preliminary decrypted text without line breaks
print("\nPreliminary decrypted text:")
print(decrypted_text, end="\n\n")


Letter | Frequency | Rank | Mapping
--------------------------------------
  C   |     37      |  1  |  E
  G   |     24      |  2  |  A
  S   |     20      |  3  |  O
  K   |     18      |  4  |  S
  Y   |     15      |  5  |  R
  I   |     15      |  6  |  D
  U   |     14      |  7  |  T
  N   |     13      |  8  |  L
  Z   |     13      |  9  |  H
  E   |     12      |  10  |  I
  O   |     10      |  11  |  N
  F   |     9      |  12  |  W
  D   |     8      |  13  |  B
  L   |     7      |  14  |  Y
  X   |     7      |  15  |  P
  J   |     7      |  16  |  C
  P   |     6      |  17  |  U
  M   |     5      |  18  |  M
  W   |     5      |  19  |  G
  H   |     5      |  20  |  F
  A   |     5      |  21  |  V
  Q   |     1      |  22  |  J

Preliminary decrypted text:
IMAYNOTBEABLETOGROWFLOWERSBUTMYGARDENPRODUCESJUSTASMANYDEADLEAVESOLDOVERSHOESPIECESOFROPEANDBUSHELSOFDEADGRASSASANYBODYSANDTODAYIBOUGHTAWHEELBARROWTOHELPINCLEARINGITUPIHAVEALWAYSLOVEDANDRESPECTEDTHEWHEELBARROWITI

**Decrypted text**

I may not be able to grow flowers but my garden produces
just as many dead leaves old overshoes pieces of rope and
bushes of dead grass as anybodys and today I bought
a wheelbarrow to help in clearing it up I have always
loved and respected the wheelbarrow it is the one
wheeled vehicle of which I am perfect master

Discoverd a word appears twice: FZCCNDGYYSF later known as → wheelbarrow
- C → e – because C is the most frequent letter
- G → A – CG assumed that e+a beacuse its also frequent
- Q → j – once
- Y → R – GAYDEN → GARDEN  
- F → W - GROF → guess: GROW
- H →  F - HLOFEY (HLOWER) → FLOWER
- U → T – That work at firts
- O → n – O occurred 5 times and is a frequent English digram.
- K → s – 4th most used letter decided to use the probability table and got it right at first
- I → d – Selected from the table
- A → v – least probability exchange from z to v
- W → g – Guessed
- M → m – IMGYNOT → I may not
The rest I used subtitution to form complete words and used the probability table.  

## - **Vigenere Cipher:**

KCCPKBGUFDPHQTYAVINRRTMVGRKDNBVFDETDGILTXRGUD
DKOTFMBPVGEGLTGCKQRACQCWDNAWCRXIZAKFTLEWRPTYC
QKYVXCHKFTPONCQQRHJVAJUWETMCMSPKQDYHJVDAHCTRL
SVSKCGCZQQDZXGSFRLSWCWSJTBHAFSIASPRJAHKJRJUMV
GKMITZHFPDISPZLVLGWTFPLKKEBDPGCEBSHCTJRWXBAFS
PEZQNRWXCVYCGAONWDDKACKAWBBIKFTIOVKCGGHJVLNHI
FFSQESVYCLACNVRWBBIREPBBVFEXOSCDYGZWPFDTKFQIY
CWHJVLNHIQIBTKHJVNPIST

**Solution** 

In [18]:
import string
from collections import Counter
import pandas as pd
from collections import defaultdict

def clean_text(text):
    """Converts text to uppercase and removes non-A-Z characters."""
    return ''.join(ch for ch in text.upper() if 'A' <= ch <= 'Z')

def char_to_num(c):
    """Converts a character (A-Z) to a number (0-25)."""
    return ord(c) - ord('A')

def num_to_char(n):
    """Converts a number (0-25) to a character (A-Z)."""
    return chr(n + ord('A'))

def index_of_coincidence(text):
    """Calculates the Index of Coincidence (IoC)."""
    if len(text) < 2:
        return 0.0
    freqs = Counter(text)
    N = len(text)
    return sum(count * (count - 1) for count in freqs.values()) / (N * (N - 1))

def average_ioc_for_key_length(ciphertext, key_length):
    """Splits text into key_length columns and computes the average IoC."""
    columns = [''] * key_length
    for i, char in enumerate(ciphertext):
        columns[i % key_length] += char
    return sum(index_of_coincidence(col) for col in columns) / key_length

def estimate_key_length_by_ioc(ciphertext, min_len=1, max_len=12):
    """Estimates key length based on IoC."""
    return sorted(((length, average_ioc_for_key_length(ciphertext, length)) 
                   for length in range(min_len, max_len + 1)), key=lambda x: x[1], reverse=True)

def find_repeated_sequences(ciphertext, min_length=3):
    """Finds repeated sequences of characters in the ciphertext."""
    sequences = defaultdict(list)
    for i in range(len(ciphertext) - min_length + 1):
        seq = ciphertext[i:i + min_length]
        sequences[seq].append(i + 1)
    return {seq: indices for seq, indices in sequences.items() if len(indices) > 1}

def find_distances(repeated_sequences):
    """Finds the distances between occurrences of repeated sequences."""
    return {seq: [indices[i + 1] - indices[i] for i in range(len(indices) - 1)] 
            for seq, indices in repeated_sequences.items()}

# Ciphertext processing
ciphertext_raw = """KCCPKBGUFDPHQTYAVINRRTMVGRKDNBVFDETDGILTXRGUD
DKOTFMBPVGEGLTGCKQRACQCWDNAWCRXIZAKFTLEWRPTYC
QKYVXCHKFTPONCQQRHJVAJUWETMCMSPKQDYHJVDAHCTRL
SVSKCGCZQQDZXGSFRLSWCWSJTBHAFSIASPRJAHKJRJUMV
GKMITZHFPDISPZLVLGWTFPLKKEBDPGCEBSHCTJRWXBAFS
PEZQNRWXCVYCGAONWDDKACKAWBBIKFTIOVKCGGHJVLNHI
FFSQESVYCLACNVRWBBIREPBBVFEXOSCDYGZWPFDTKFQIY
CWHJVLNHIQIBTKHJVNPIST""".replace("\n", "")

ciphertext = clean_text(ciphertext_raw)

# Find repeated sequences
repeated_sequences = find_repeated_sequences(ciphertext)
distances = find_distances(repeated_sequences)
df_sequences = pd.DataFrame([(seq, indices, distances[seq]) 
                             for seq, indices in repeated_sequences.items()],
                            columns=['Sequence', 'Positions', 'Intervals'])

# Estimate key length using IoC
ioc_results = estimate_key_length_by_ioc(ciphertext)
df_ioc = pd.DataFrame(ioc_results, columns=['Key Length', 'Average IoC'])

# Display results
print("Repeated sequences and their distances:")
print(df_sequences.to_string(index=False))

print("\nKey length estimates based on IoC:")
print(df_ioc.to_string(index=False))


Repeated sequences and their distances:
Sequence                 Positions         Intervals
     MVG                 [23, 179]             [156]
     BVF                 [30, 294]             [264]
     DDK                 [45, 243]             [198]
     KFT             [80, 98, 254]         [18, 156]
     HJV [108, 126, 264, 318, 330] [18, 138, 54, 12]
     HCT                [131, 215]              [84]
     RLS                [134, 152]              [18]
     KCG                [139, 260]             [121]
     AFS                [163, 223]              [60]
     RWX                [219, 231]              [12]
     VYC                [235, 277]              [42]
     WBB                [250, 286]              [36]
     BBI                [251, 287]              [36]
     JVL                [265, 319]              [54]
     VLN                [266, 320]              [54]
     LNH                [267, 321]              [54]
     NHI                [268, 322]              [54]

Key l

First, we applied the Kasiski Test, and we found that the trigram HJV occurs five times at positions 108, 126, 264, 318, and 330. We also noticed that the intervals are 18, 138, 54, and 12, which are divisible by 2, 3, and 6. So we assumed that the keyword length is either 2, 3 or 6. Based on this supposition, we computed the Index of Coincidence, which specified that the most likely key length was indeed 6.

In [20]:
def shifter_decrypt(char, shift):
    """Decrypts a character by applying an inverse Caesar shift."""
    return num_to_char((char_to_num(char) - shift) % 26)

def find_best_shift_for_column(col, language='EN'):
    """Finds the best shift using chi-squared frequency analysis."""
    freq_en = {'A': 0.08167, 'B': 0.01492, 'C': 0.02782, 'D': 0.04253, 'E': 0.12702,
               'F': 0.02228, 'G': 0.02015, 'H': 0.06094, 'I': 0.06966, 'J': 0.00153,
               'K': 0.00772, 'L': 0.04025, 'M': 0.02406, 'N': 0.06749, 'O': 0.07507,
               'P': 0.01929, 'Q': 0.00095, 'R': 0.05987, 'S': 0.06327, 'T': 0.09056,
               'U': 0.02758, 'V': 0.00978, 'W': 0.02360, 'X': 0.00150, 'Y': 0.01974,
               'Z': 0.00074}
    
    def chi_squared(decrypted_col):
        dec_counts = Counter(decrypted_col)
        return sum(((dec_counts[letter] - freq_en[letter] * len(col)) ** 2) /
                   (freq_en[letter] * len(col) + 1e-9) for letter in string.ascii_uppercase)
    
    return min(range(26), key=lambda shift: chi_squared(''.join(shifter_decrypt(ch, shift) for ch in col)))

def guess_key_by_frequency(ciphertext, key_length):
    """Finds the key for Vigenère cipher using frequency analysis."""
    columns = [''] * key_length
    for i, char in enumerate(ciphertext):
        columns[i % key_length] += char
    return ''.join(num_to_char(find_best_shift_for_column(col)) for col in columns)

def vigenere_decrypt(ciphertext, key):
    """Decrypts a Vigenère cipher text using the given key."""
    key_length = len(key)
    return ''.join(num_to_char((char_to_num(c) - char_to_num(key[i % key_length])) % 26) 
                   for i, c in enumerate(ciphertext))

# Extract probable key length
probable_length = ioc_results[0][0]
guessed_key = guess_key_by_frequency(ciphertext, probable_length)

# Display results
print(f"Likely key length: {probable_length}")
print(f"Estimated key: {guessed_key}")
print("\nDecrypted text:")
print(vigenere_decrypt(ciphertext, guessed_key))


Likely key length: 6
Estimated key: CRYPTO

Decrypted text:
ILEARNEDHOWTOCALCULATETHEAMOUNTOFPAPERNEEDEDFORAROOMWHENIWASATSCHOOLYOUMULTIPLYTHESQUAREFOOTAGEOFTHEWALLSBYTHECUBICCONTENTSOFTHEFLOORANDCEILINGCOMBINEDANDDOUBLEITYOUTHENALLOWHALFTHETOTALFOROPENINGSSUCHASWINDOWSANDDOORSTHENYOUALLOWTHEOTHERHALFFORMATCHINGTHEPATTERNTHENYOUDOUBLETHEWHOLETHINGAGAINTOGIVEAMARGINOFERRORANDTHENYOUORDERTHEPAPER


As mentioned above, we used the Mg method, with the following formula: 
$$
M_g = \sum_{i=0}^{25} \frac{p_i f_{i+g}}{n'}
$$
 and we found relative shifts. Specifically, we noticed that the the keyword is the result of some shift applied to APWNRM. After analyzing each of them, we concluded that the only one that made sense was CRYPTO (or A |-> C). Finally, once we found the keyword, we inverse produced it to the plaintext. 

**Decrypted text**

I learned how to calculate the amount of paper needed for a room when I was at school. You multiply the square footage of the walls by the cubic contents of the floor and ceiling combined and double it. You then allow half the total for openings such as windows and doors. Then you allow the other half for matching the pattern. Then you double the whole thing again to give a margin of error, and then you order the paper.

## - **Affine Cipher:**

KQEREJEBCPPCJCRKIEACUZBKRVPKRBCIBQCARBJCVFCUP
KRIOFKPACUZQEPBKRXPEIIEABDKPBCPFCDCCAFIEABDKP
BCPFEQPKAZBKRHAIBKAPCCIBURCCDKDCCJCIDFUIXPAFF
ERBICZDFKABICBBENEFCUPJCVKABPCYDCCDPKBCOCPERK
IVKSCPICBRKIJPKABI

**Solution** 

In [15]:
from collections import Counter
import pandas as pd

def analyze_ciphertext(ciphertext):
    # Eliminar espacios y saltos de línea
    cleaned_text = ciphertext.replace("\n", "").replace(" ", "")
    
    # Contar frecuencia de cada letra
from collections import Counter
import pandas as pd

def analyze_ciphertext(ciphertext):
    # Eliminar espacios y saltos de línea
    cleaned_text = ciphertext.replace("\n", "").replace(" ", "")
    
    # Contar frecuencia de cada letra
    frequency = Counter(cleaned_text)
    
    # Ordenar por frecuencia en orden descendente
    sorted_freq = sorted(frequency.items(), key=lambda x: x[1], reverse=True)
    
    # Asignar rank basado en frecuencia
    rank = {letter: rank+1 for rank, (letter, _) in enumerate(sorted_freq)}
    
    # Crear DataFrame para visualización
    df = pd.DataFrame(sorted_freq, columns=['Letter', 'Frequency'])
    df['Rank'] = df['Frequency'].rank(method='min', ascending=False).astype(int)
    
    return df

# Texto cifrado
ciphertext = """KQEREJEBCPPCJCRKIEACUZBKRVPKRBCIBQCARBJCVFCUP
KRIOFKPACUZQEPBKRXPEIIEABDKPBCPFCDCCAFIEABDKP
BCPFEQPKAZBKRHAIBKAPCCIBURCCDKDCCJCIDFUIXPAFF
ERBICZDFKABICBBENEFCUPJCVKABPCYDCCDPKBCOCPERK
IVKSCPICBRKIJPKABI"""

# Análisis de frecuencia
df_result = analyze_ciphertext(ciphertext)

# Mostrar resultado
print(df_result)

   Letter  Frequency  Rank
0       C         32     1
1       B         21     2
2       K         20     3
3       P         20     3
4       I         16     5
5       E         13     6
6       A         13     6
7       R         12     8
8       F         10     9
9       D          9    10
10      J          6    11
11      U          6    11
12      Q          4    13
13      Z          4    13
14      V          4    13
15      O          2    16
16      X          2    16
17      H          1    18
18      N          1    18
19      Y          1    18
20      S          1    18


We first analyze the frequency of letters in the ciphertext. By comparing these frequencies with English letter frequencies, we can attempt to deduce common letters like E, T, and A.
We can assume: 
- C → E
- B → T

We select two common letters in the ciphertext and assume their corresponding plaintext values.

Solve for 𝑎 and 𝑏 using the equation:

$$
E(x_1) = (a x_1 + b) \mod 26
$$

$$
E(x_2) = (a x_2 + b) \mod 26
$$

- E position on the alphabet is 4 = $ x_1 $ 
- T position on the alphabet is 19 = $ x_2 $
- First we find a and b and then compute:

$$
D(y) = a^{-1} (y - b) \mod 26
$$

Se utiliza:

- Affine Cipher Equations

**Encryption:**
$$
e_k(x) = a x + b
$$

**Decryption:**
$$
d_k(y) = a^{-1} (y - b)
$$


In [20]:
from collections import Counter
import pandas as pd
import math

def mod_inverse(a, m):
    # Find the modular inverse of 'a' mod 'm'
    for x in range(1, m):
        if (a * x) % m == 1:
            return x
    return None  # Return None if no modular inverse exists

def affine_decrypt(ciphertext, a, b):
    # Remove spaces and line breaks from the ciphertext
    cleaned_text = ciphertext.replace("\n", "").replace(" ", "")
    
    # Find the modular inverse of 'a' mod 26
    a_inv = mod_inverse(a, 26)
    if a_inv is None:
        raise ValueError("No modular inverse for the given 'a'. Choose a different value.")
    
    # Decryption function
    decrypted_text = ""
    for char in cleaned_text:
        if char.isalpha():
            y = ord(char) - ord('A')  # Convert letter to number (A=0, B=1, ..., Z=25)
            x = (a_inv * (y - b)) % 26  # Apply decryption formula
            decrypted_text += chr(x + ord('A'))  # Convert number back to letter
        else:
            decrypted_text += char  # Keep non-alphabetic characters unchanged
    
    return decrypted_text

# Encrypted text (ciphertext)
ciphertext = """KQEREJEBCPPCJCRKIEACUZBKRVPKRBCIBQCARBJCVFCUP
KRIOFKPACUZQEPBKRXPEIIEABDKPBCPFCDCCAFIEABDKP
BCPFEQPKAZBKRHAIBKAPCCIBURCCDKDCCJCIDFUIXPAFF
ERBICZDFKABICBBENEFCUPJCVKABPCYDCCDPKBCOCPERK
IVKSCPICBRKIJPKABI"""

# Affine Cipher parameters
a = 19
b = 4

# Decrypt the text
decrypted_text = affine_decrypt(ciphertext, a, b)

# Display result
print("Decrypted text:")
print(decrypted_text)

Decrypted text:
OCANADATERREDENOSAIEUXTONFRONTESTCEINTDEFLEURONSGLORIEUXCARTONBRASSAITPORTERLEPEEILSAITPORTERLACROIXTONHISTOIREESTUNEEPOPEEDESPLUSBRILLANTSEXPLOITSETTAVALEURDEFOITREMPEEPROTEGERANOSFOYERSETNOSDROITS


**Decrypted text**

O CANADA TERRE DE NOS AIEUX TON FRONT EST CEINT DE FLEURONS GLORIEUX CAR TON BRAS SAIT PORTER L EPEE IL SAIT PORTER LA CROIX TON HISTOIRE EST UNE EPOPEE DES PLUS BRILLANTS EXPLOITS ET TA VALEUR DE FOI TREMPEE PROTEGERA NOS FOYERS ET NOS DROITS


## - **Other cipher:**

**Solution** 

As we were only given the ciphertext, we had to work with it in order to infer the cipher algorithm we were dealing with. 

In [1]:
import string
from collections import Counter

# --------------------------------------------------
# 1. Cleanup and utilities
# --------------------------------------------------

def clean_text(text):
    """
    Converts the text to uppercase and removes anything 
    that is not an A-Z letter.
    """
    text = text.upper()
    cleaned = []
    for ch in text:
        if 'A' <= ch <= 'Z':
            cleaned.append(ch)
    return "".join(cleaned)

def char_to_num(c):
    """
    Converts a character (A-Z) to a number (0-25).
    """
    return ord(c) - ord('A')

def num_to_char(n):
    """
    Converts a number (0-25) to a character (A-Z).
    """
    return chr(n + ord('A'))

So, we started analyzing the **Index of Coincidence**, a statistical measure that estimates how likely it is for two randomly chosen letters from a text to be identical.

In natural language texts (like English), certain letters (such as E, T, A) appear more frequently than others. This produces an IoC typically around 0.065 to 0.07.

For a monoalphabetic substitution cipher, the IoC tends to be close to that of the original language (around 0.065 for English). For completely random text, the IoC is much lower (approximately 0.038).

We needed to find the IoC for the given ciphertext, so we could conclude if we were dealing with a monoalphabetic (substitution) or polyalphabetic cipher like Vigenère.

In [2]:
# --------------------------------------------------
# 2. Calculation of the Index of Coincidence (IoC)
# --------------------------------------------------

def index_of_coincidence(text):
    """
    Calculates the Index of Coincidence (IoC) for a text
    in uppercase (A-Z).
    
    Formula: IoC = sum( f_i * (f_i - 1) ) / [ N * (N - 1) ]
    where f_i is the frequency of each letter,
    and N is the total length of the text.
    """
    if len(text) < 2:
        return 0.0
    
    freqs = Counter(text)
    N = len(text)
    numerator = 0
    for letter, count in freqs.items():
        numerator += count * (count - 1)
    denominator = N * (N - 1)
    return numerator / denominator

When we executed this function, we found that the ciphertext was encoded with a polyalphabetic.

As the next step, we defined a function that calculated an estimated key length, sp we could split the ciphertext into that many columns, each resembling a simple Caesar cipher of natural language.

In [3]:
def average_ioc_for_key_length(ciphertext, key_length):
    """
    Splits the text into 'key_length' columns and computes
    the average IoC of those columns.
    """
    columns = [''] * key_length
    
    # Distribute the text into columns
    for i, char in enumerate(ciphertext):
        columns[i % key_length] += char
    
    # Calculate the IoC of each column and compute the average
    ioc_sum = 0
    for col in columns:
        ioc_sum += index_of_coincidence(col)
    return ioc_sum / key_length

def estimate_key_length_by_ioc(ciphertext, min_len=1, max_len=12):
    """
    Tries key lengths from min_len to max_len,
    calculates the average IoC, and returns a list
    sorted by IoC (descending).
    """
    results = []
    for length in range(min_len, max_len+1):
        avg_ioc = average_ioc_for_key_length(ciphertext, length)
        results.append((length, avg_ioc))
    
    # Sort by descending IoC
    results.sort(key=lambda x: x[1], reverse=True)
    return results

We also defined a function to decrypt each column.

In [4]:
def caesar_decrypt(char, shift):
    """
    Decrypts a character (A-Z) by applying an inverse
    Caesar shift of 'shift'.
    """
    return num_to_char((char_to_num(char) - shift) % 26)

By testing all 26 possible shifts on a column and calculating the chi-squared value for each candidate decryption, the shift that minimizes the chi-squared statistic is likely the correct one. The chi-squared test has been a standard tool in cryptanalysis, particularly for breaking classical ciphers like Caesar or Vigenère.

This is because a lower chi-squared value indicates that the observed frequencies closely match the natural language distribution.

In [5]:
def find_best_shift_for_column(col, language='EN'):
    """
    Finds the Caesar shift that makes the column 'col'
    closest to the typical frequency distribution 
    (English or Spanish).
    """
    # Approximate frequencies for English (you can adjust if you want Spanish)
    freq_en = {
        'A': 0.08167, 'B': 0.01492, 'C': 0.02782, 'D': 0.04253, 'E': 0.12702,
        'F': 0.02228, 'G': 0.02015, 'H': 0.06094, 'I': 0.06966, 'J': 0.00153,
        'K': 0.00772, 'L': 0.04025, 'M': 0.02406, 'N': 0.06749, 'O': 0.07507,
        'P': 0.01929, 'Q': 0.00095, 'R': 0.05987, 'S': 0.06327, 'T': 0.09056,
        'U': 0.02758, 'V': 0.00978, 'W': 0.02360, 'X': 0.00150, 'Y': 0.01974,
        'Z': 0.00074
    }
    
    # Count the actual frequency of the column
    counts = Counter(col)
    length = len(col)
    
    def chi_squared(decrypted_col):
        # Computes the chi-squared against the 'freq_en' distribution
        c2 = 0.0
        dec_counts = Counter(decrypted_col)
        for letter in string.ascii_uppercase:
            observed = dec_counts[letter]
            expected = freq_en[letter] * len(decrypted_col)
            c2 += (observed - expected)**2 / (expected + 1e-9)
        return c2
    
    best_shift = 0
    min_chi = float('inf')
    for shift in range(26):
        # Decrypt the column with this shift
        decrypted_col = ''.join(caesar_decrypt(ch, shift) for ch in col)
        c2 = chi_squared(decrypted_col)
        if c2 < min_chi:
            min_chi = c2
            best_shift = shift
    
    return best_shift

For each column, all possible shifts are evaluated using the chi-squared test, the one that produces the letter frequency distribution most similar to natural language is selected.

The selected shifts for each column are converted into letters. This sequence of letters forms the candidate key.

In [6]:
def guess_key_by_frequency(ciphertext, key_length, language='EN'):
    """
    Splits the ciphertext into 'key_length' columns and finds 
    the best shift (shift) for each column.
    Returns the key as a string of letters (A-Z).
    """
    columns = [''] * key_length
    for i, char in enumerate(ciphertext):
        columns[i % key_length] += char
    
    # For each column, we compute the "optimal" shift
    key_shifts = []
    for col in columns:
        shift = find_best_shift_for_column(col, language=language)
        key_shifts.append(shift)
    
    # Convert the shifts to letters of the key
    # Remember that in Vigenère, shift=0 -> 'A', shift=1 -> 'B', etc.
    key = ''.join(num_to_char(s) for s in key_shifts)
    return key

Once the key is guessed, it is applied to the entire ciphertext using the Vigenère decryption method. If the output is readable and makes sense, the key is likely correct.

In [8]:
def vigenere_decrypt(ciphertext, key):
    """
    Decrypts a 'ciphertext' (A-Z) with the 'key' (A-Z)
    using the classic Vigenère cipher.
    """
    plaintext = []
    key_index = 0
    key_length = len(key)
    
    for c in ciphertext:
        shift = char_to_num(key[key_index])
        # Inverse shift
        p = (char_to_num(c) - shift) % 26
        plaintext.append(num_to_char(p))
        key_index = (key_index + 1) % key_length
    
    return "".join(plaintext)

Lastly, we tried to break the ciphertext with the explained method.

In [9]:
if __name__ == "__main__":
    ciphertext_raw = """
    BNVSNSIHQCEELSSKKYERIFJKXUMBGYKAMQLJTYAVFBKVT
DVBPVVRJYYLAOKYMPQSCGDLFSRLLPROYGESEBUUALRWXM
MASAZLGLEDFJBZAVVPXWICGJXASCBYEHOSNMULKCEAHTQ
OKMFLEBKFXLRRFDTZXCIWBJSICBGAWDVYDHAVFJXZIBKC
GJIWEAHTTOEWTUHKRQVVRGZBXYIREMMASCSPBNLHJMBLR
FFJELHWEYLWISTFVVYFJCMHYUYRUFSFMGESIGRLWALSWM
NUHSIMYYITCCQPZSICEHBCCMZFEGVJYOCDEMMPGHVAAUM
ELCMOEHVLTIPSUYILVGFLMVWDVYDBTHFRAYISYSGKVSUU
HYHGGCKTMBLRX
    """.strip()
    
    # 1) Clean the text
    ciphertext = clean_text(ciphertext_raw)
    
    # 2) Estimate key lengths with IoC
    ioc_results = estimate_key_length_by_ioc(ciphertext, min_len=1, max_len=12)
    print("Possible key lengths (sorted by IoC):")
    for length, ioc_val in ioc_results:
        print(f"  Length = {length}, Average IoC = {ioc_val:.4f}")
    
    # Assuming we pick the key length with the highest IoC (e.g., 6)
    probable_length = ioc_results[0][0]
    print(f"\nBased on IoC, a likely key length is: {probable_length}")
    
    # 3) Try to guess the key via frequency analysis
    guessed_key = guess_key_by_frequency(ciphertext, probable_length, language='EN')
    print(f"Estimated key (frequencies): {guessed_key}")
    
    # 4) Decrypt with that key
    trial_plaintext = vigenere_decrypt(ciphertext, guessed_key)
    print("\nTentative plaintext:")
    print(trial_plaintext)

Possible key lengths (sorted by IoC):
  Length = 12, Average IoC = 0.0648
  Length = 6, Average IoC = 0.0606
  Length = 8, Average IoC = 0.0513
  Length = 4, Average IoC = 0.0486
  Length = 3, Average IoC = 0.0468
  Length = 10, Average IoC = 0.0463
  Length = 2, Average IoC = 0.0453
  Length = 9, Average IoC = 0.0445
  Length = 11, Average IoC = 0.0424
  Length = 7, Average IoC = 0.0418
  Length = 5, Average IoC = 0.0415
  Length = 1, Average IoC = 0.0414

Based on IoC, a likely key length is: 12
Estimated key (frequencies): THEORYTHEORY

Tentative plaintext:
IGREWUPAMONGSLOWTALKERSMENINPARTICULARWHODROPPEDWORDSAFEWATATIMELIKEBEANSINAHILLANDWHENIGOTTOMINNEAPOLISWHEREPEOPLETOOKALAKEWOBEGONCOMMATOMEANTHEENDOFASTORYICOULDNTSPEAKAWHOLESENTENCEINCOMPANYANDWASCONSIDEREDNOTTOOBRIGHTSOIENROLLEDINASPEECHCOURSETAUGHTBYORVILLESANDTHEFOUNDEROFREFLEXIVERELAXOLOGYASELFHYPNOTICTECHNIQUETHATENABLEDAPERSONTOSPEAKUPTOTHREEHUNDREDWORDSPERMINUTE


I grew up among slow talkers. Men in particular who dropped words a few at a time like beans in a hill. And when I got to Minneapolis, where people took a Lake Wobegon comma to mean the end of a story, I couldn't speak a whole sentence in company and was considered not too bright. So I enrolled in a speech course taught by Orville Sand, the founder of reflexive relaxology, a self hypnotic technique that enabled a person to speak up to three hundred words per minute.

## **4. Prove that the Affine Cipher achieves perfect secrecy if every key is used with equal probability 1/312.**

For each pair of plaintext-ciphertext letters $(x,y)$, there are 12 keys that encrypt $x$ to $y$:

For each choice of $a$, the key $(a,y−ax)$ encrypts the plaintext letter $x$ to the ciphertext letter $y$, since $ax+ (y−ax) = y$. We know that there are 12 possible chocies for $a$ since $a$ must be chosen such that $gcd(a,26)=1$, and there are only twelve numbers that satisfy this condition ${1,3,5,7,9,11,15,17,19,21,23,25}$. So, there are exactly twelve keys that map a given plaintext letter to a given ciphertext letter. 

Thus:
$$
p_c(y) = \sum_{k \in K} p_K(k) p_P(d_k(y)) = \frac{12}{312} p_P(a) + \frac{12}{312} p_P(b) + \cdots + \frac{12}{312} p_P(z) = \frac{12}{312} \times 1 = \frac{1}{26}.
$$

Taking into account that:
$$
p_C(y | x) = \sum_{k : x = d_k(y)} p_K k = \frac{12}{312} = \frac{1}{26},
$$ 

each key occurs with a probability of $ \frac{1}{312} $ and there are twelve keys that decrypt $ x $ to $ y $, which means that there are twelve keys for which $ x = d_k(y) $.

Hence
$$
p_P(x | y) = \frac{p_P(x) p_C(y | x)}{p_C(y)} = \frac{p_P(x) \frac{1}{26}}{\frac{1}{26}} = p_P(x)
$$

which proves that this cryptosystem has perfect secrecy.





## **5. Prove that if a cryptosystem has perfect secrecy and |K|= |C|= |P|, then every ciphertext is equally probable.**

Given that the cryptosystm has perfect secrecy, then it means that:

$$
\forall p \in P, c \in C, Pr_{P,C}(p / c) = Pr_P(p)
$$

So, using the Bayes Theorem we can deduce that :

$$
\forall c \in C, Pr(c) = \frac{Pr(p) \cdot Pr(c / p)}{Pr(p / c)} = \frac{Pr(p) \cdot Pr(c / p)}{Pr(p)} = Pr(c / p)
$$

As $\ |P| = |C| = |K| \ $, we know that there is only one key among $n$ that encrypts $p$ to $c$. So

$$
\forall c \in C, Pr(c / p) = \frac{1}{n}
$$

We can conclude that every ciphertext is equally probable.


## **6. Use the EXTENDED EUCLIDEAN ALGORITHM to compute the following multiplicative inverses:**

![Parte1](imgs/6parte1.jpeg)

![Parte1](imgs/6parte2.jpeg)

![Parte1](imgs/6parte3.jpeg)